DRAFT-PR: change zfLogin logic with tokenPool design

MucheXD opened the pull request:
这是一个草稿 PR

这个 PR 初步完成了使用令牌池的登录逻辑，现行逻辑如下：

1. main 首先启动 `loginTokenManager` 并初始化一个实例
2. `loginTokenManager` 根据预设的参数不断请求新的 loginToken 并加入池中
3. login 函数从 loginToken 池拿取一个并直接发起最终请求

包含以下较大变动：

1. 加入了用于实现池维护的模块
2. 修改了验证码破解逻辑，现在验证码将使用从本地读取模板并对比的方式，另有备用破解逻辑使用 Lab 颜色空间比对色值
3. 原有的 login 函数从发起完整请求逻辑更改为从 loginToken 池拿取一个并直接发起最终请求

由于以下原因，这个草稿 PR 无法正式发起合并：

~~1. 根据计划，应该将 tokenPool 存储位置更改为 Redis 服务~~ (Finished)
2. 代码仅仅测试了能否运行，暂未运行条件测试和压力测试 (WIP)
~~3. 代码中仍包含调试用输出语句~~ (->4)
4. 代码中错误处理部分并没有统一， 需要设计错误侦测与日志输出 [REV](https://github.com/zjutjh/funnel/pull/41#discussion_r2478828959)
5. 代码中仍含有无效的旧逻辑片段（包括注释的部分）[REV](https://github.com/zjutjh/funnel/pull/41#discussion_r2478777226)
6. 统一 resty 客户端初始化方式 [REV](https://github.com/zjutjh/funnel/pull/41#discussion_r2478770814)

请审查并指出其它未于上述列出的问题 (●'◡'●)

MucheXD committed:
feat: import captcha-cracker & token-rigister

chores: add redis connection check; add [type] key to CheckLoginForm as need; enable cspell;

MucheXD committed:
feat: add pool manager

MucheXD committed:
feat: changed login logic using loginTkPool

res: add captcha_templates with 10 pics

cbluebird commented on app/service/zfService/dataGen.go:
```diff
@@ -1,11 +1,16 @@
 package zfService
 
 import (
+	"crypto/rand"
```
lint

cbluebird commented on app/service/zfService/dataGen.go:
```diff
@@ -33,14 +38,30 @@ func genTermRelatedInfoReqData(year string, term string) url.Values {
+	n, _ := new(big.Int).SetString(hex.EncodeToString(nString), 16)
+	eString, _ := base64.StdEncoding.DecodeString(exponent)
+	e, _ := strconv.ParseInt(hex.EncodeToString(eString), 16, 32)
+	pub := rsa.PublicKey{E: int(e), N: n}
```
err 处理

cbluebird replied:
顺便加个注释

MucheXD replied:
可恶啊，这一坨是原先代码就有的，拿过来用用成我负责的了 ヾ(≧へ≦)〃

cbluebird commented on app/service/zfService/login.go:
```diff
@@ -21,66 +23,140 @@ type captchaServerResponse struct {
-	//		config.Redis.Set("zf_url", "new", 0)
-	//	}
-	//}
+	ltk, err := loginTokenManager.LoginTkMgr.GetAToken()
```
这是啥函数名，不要缩写

MucheXD replied:
太长的函数名好像又不符合规范来着...？原先特意改短的，现在我改回去xd

cbluebird commented on app/service/zfService/login.go:
```diff
@@ -21,66 +23,140 @@ type captchaServerResponse struct {
-		if v.Name == "JSESSIONID" {
-			sessionCookie = v
+	// 构造 resty 客户端
+	client := resty.New().
+		SetTimeout(5 * time.Second).
+		SetRedirectPolicy(resty.NoRedirectPolicy()) // 禁止自动重定向
```
需要每个函数构建一个吗？是否可以用一个
通过依赖注入来解耦？

MucheXD replied:
整个 funnel 只有我提交的部分使用了 resty ，所以这里没有全局的 resty 对象。这项目一半原生 http 一半 resty 的看的确实也很难受，但是要全改恐怕工作量有点大。（改为 resty 的原因是我们认为原来的 fetch 模块写的确实依托）

cbluebird commented on app/service/zfService/login.go:
```diff
@@ -21,66 +23,140 @@ type captchaServerResponse struct {
+// 	f.Init()
+// 	_, err := f.Get(zf.ZfLoginHome())
+// 	//if strings.Contains(err.Error(), "context deadline exceeded") {
+// 	//	if strings.Compare(config.Redis.Get("zf_url").String(), "new") == 0 {
```
直接删

cbluebird commented on app/service/zfService/loginTokenManager/loginTokenManager.go:
```diff
@@ -0,0 +1,139 @@
+	go m.runPoolRoutineMt()
+}
+
+func (m *LoginTokenManager) GetAToken() (*LoginToken, error) {
```
不要这么命名

cbluebird commented on app/service/zfService/loginTokenManager/loginTokenManager.go:
```diff
@@ -0,0 +1,139 @@
+}
+
+func (m *LoginTokenManager) Init(host string) {
+	m.tokenPool = &loginTokenPool{}
+	m.hostUrl = host
+	m.preferredTokenCount = 50
+	m.maxResisters = 10
+	m.currentRunningResisters = 0
```
需要这么写吗？这么写好看点吧
&login{
token: 1,
}

cbluebird replied:
并且都是构造函数了，为什么还要套一层Init

MucheXD replied:
两个构造函数作用不同。

这个文件混合了两区代码，其一实现具体的 `TokenManager` 逻辑并封装成一个结构体，另一则初始化这个结构体并将一个实例赋值给包级变量方便外部调用。

如果将后者拆分，会增加一个仅有几行代码的文件。需要这样做吗？

cbluebird commented on app/service/zfService/loginTokenManager/loginTokenManager.go:
```diff
@@ -0,0 +1,139 @@
+		var pfRegister int32
+		if currentPoolSize > int32(float32(m.preferredTokenCount)*0.25) { // 池填充 25% 以上
+			// 启动部分注册器 根据池的当前大小动态调整
+			pfRegister = int32(float32(m.maxResisters) * (1 - float32(currentPoolSize)/float32(m.preferredTokenCount)))
```
为什么你这么喜欢类型转换啊，能不能统一下精度和单位？

MucheXD replied:
此处的类型转换的目的是进行小数运算。如果你不喜欢，我可以改成 ×100 再 /100 的形式来完成这一任务，但那样做我认为不太直观。

cbluebird commented on config/redis.go:
```diff
@@ -17,4 +19,10 @@ func init() {
+	Redis.Set("CONNECTION_TEST", "CONNECTION_TEST", time.Second)
+	redisVal, err := Redis.Get("CONNECTION_TEST").Result()
+	if err != nil || redisVal != "CONNECTION_TEST" {
+		panic("Redis connection failed")
```
这啥呀

MucheXD replied:
用于验证Redis服务是否正确连接的逻辑。如果不需要，可以移除。

cbluebird commented on app/service/zfService/login.go:
```diff
@@ -1,18 +1,20 @@
 	"strings"
+	"time"
+
+	"github.com/PuerkitoBio/goquery"
```
这个包干啥用的

MucheXD replied:
这个不是我使用的包，它用于解析html并查询相关字段。在我的代码里使用正则表达式来进行相关操作。显示我增加了此行，可能是格式化工具造成的导入顺序变动。

cbluebird commented on app/service/zfService/loginTokenManager/runRegister.go:
```diff
@@ -0,0 +1,70 @@
+	}
+
+	// 验证码未通过
+	return LoginToken{}, fmt.Errorf("captcha not passed")
```
error看一下怎么处理？

cbluebird replied:
打个日志？

MucheXD replied:
我翻了 funnel 的代码，没有关于 log 的标准用法。如果没有指定，我将使用原生的 log 并硬编码消息。

cbluebird commented on app/service/zfService/zfCaptchaCracker/analyzeCracker.go:
```diff
@@ -0,0 +1,215 @@
+	L1, a1, b1 := c.labConverter.RgbToLab(c1)
+	L2, a2, b2 := c.labConverter.RgbToLab(c2)
+
+	if log {
```
这是啥呀

MucheXD replied:
关于 analyzeCracker ，其中涉及到颜色空间转换的运算全部为 AI 生成的代码。这个 cracker 主要负责二层验证码破解，即首选破解器失效时将自动尝试该破解器。之所以设计两层，是因为首层的 compareCracker 有其特殊性，只能针对本地预置的图片进行破解，虽然效率很高，但考虑到某天可能更换验证码图片，故设计了二层破解器。二层破解器的效率较低，成功率只有约 90% ，如果多次发生降级，可能要发出警报。另外，我正在尝试（不一定落实）将原有的使用远程破解器的方案整合为第三层破解器，以免验证码拼图图形发生变化。

MucheXD replied:
`analyzeCracker`现已重构

cbluebird commented on app/service/zfService/zfCaptchaCracker/analyzeCracker.go:
```diff
@@ -0,0 +1,215 @@
+		}
+	}
+	return 0, fmt.Errorf("targeted left edge not found") // ERR_HANDLE_8
+}
```
错误日志都注意下呗

cbluebird commented on cspell.json:
这个又是什么

MucheXD replied:
提供自动拼写更正插件使用的自定义词库。由于前端团队已经开始使用类似的方案，所以没有添加 `.gitignore`，如果我们不需要，我会在下一次提交移除。

cbluebird commented:
整体逻辑太复杂了，tokenPool不要放Service了吧，这和你的业务逻辑没什么关系，放在pkg里吧，相当于一个独立的组建。给biz提供一些支撑～

MucheXD committed:
feat: token-pool now using redis

MucheXD commented:
> 整体逻辑太复杂了，tokenPool不要放Service了吧，这和你的业务逻辑没什么关系，放在pkg里吧，相当于一个独立的组建。给biz提供一些支撑～

tokenPool 涉及到存取 LoginToken，这个对象是业务特有的。如果需要做成通用组件，则可能要考虑使用接口。

SugarMGP commented on captcha_templates/1.png:
beautiful view

MucheXD committed:
chore: several changes to meet requirements

1. feat: refactoring analyseCracker, useing more readble method to judge left-edge and improve acc-rate.

2. chore: remove debugging log outputs

3. chore: remove unused old codes

4. style: change name for some vars to make it more readable

5. chore: add more comments to make codes more understandable

6. chore: remove debugging codes

7. chore: add error handling for some functions

MangoGovo commented:
commit 说中文
